### PR TITLE
CLI Tabulation and Staker UI

### PR DESCRIPTION
**Pull request by KPrasch**
Advances #1621 (full fix after #1652)
```
$ nucypher stake list
 ____    __            __                      
/\  _`\ /\ \__        /\ \                     
\ \,\L\_\ \ ,_\    __ \ \ \/'\      __   _ __  
 \/_\__ \\ \ \/  /'__`\\ \ , <    /'__`\/\`'__\
   /\ \L\ \ \ \_/\ \L\.\\ \ \\`\ /\  __/\ \ \/ 
   \ `\____\ \__\ \__/.\_\ \_\ \_\ \____\\ \_\ 
    \/_____/\/__/\/__/\/_/\/_/\/_/\/____/ \/_/ 
   
The Holder of Stakes.                                      


Staker 0xDCc3Af9Ad1e23047156e92621D8FFC7C1da3b6eC ════
Worker NO_WORKER_ASSIGNED ════
--------------  ----------------------
Status          Missing 1 confirmation
Restaking       Yes (Unlocked)
Winding Down    No
Unclaimed Fees  0 Gwei
--------------  ----------------------
╒═══════╤══════════╤═════════════╤══════════════════╕
│   Idx │ Value    │   Remaining │ Enactment        │
╞═══════╪══════════╪═════════════╪══════════════════╡
│     0 │ 45000 NU │         366 │ Jan 26 16:00 PST │
╘═══════╧══════════╧═════════════╧══════════════════╛

Staker 0xcC0678E51a8237b762c09d6548d2d07285609e98 ════
Worker 0xe1887a501647BaD48e58fE62e6165D83EC817c79 ════
--------------  ----------------
Status          Confirmed #18288
Restaking       No (Unlocked)
Winding Down    No
Unclaimed Fees  3E-9 Gwei
--------------  ----------------
╒═══════╤════════════╤═════════════╤══════════════════╕
│   Idx │ Value      │   Remaining │ Enactment        │
╞═══════╪════════════╪═════════════╪══════════════════╡
│     0 │ 1000000 NU │         367 │ Jan 20 16:00 PST │
╘═══════╧════════════╧═════════════╧══════════════════╛
(nucypher) kieran@pegasus:~/Git/nucypher$ 

```

```
$ nucypher stake accounts

/\  _`\ /\ \__        /\ \                     
\ \,\L\_\ \ ,_\    __ \ \ \/'\      __   _ __  
 \/_\__ \\ \ \/  /'__`\\ \ , <    /'__`\/\`'__\
   /\ \L\ \ \ \_/\ \L\.\\ \ \\`\ /\  __/\ \ \/ 
   \ `\____\ \__\ \__/.\_\ \_\ \_\ \____\\ \_\ 
    \/_____/\/__/\/__/\/_/\/_/\/_/\/____/ \/_/ 
   
The Holder of Stakes.                                      

╒═══════════╤════════════════════════════════════════════╤══════════════════════════╤══════╕
│ Staking   │ Account                                    │ ETH                      │ NU   │
╞═══════════╪════════════════════════════════════════════╪══════════════════════════╪══════╡
│ No        │ 0x4F1D703D0d0D424E45b760A3AE9E06027Ba4D4AC │ 0.099987574350439578 ETH │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ No        │ 0x66Afba11abC8d50eC7e60531De90870920cCBEf0 │ 0 ETH                    │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ No        │ 0xBB45F532414418088463A0680C62Ec3A9804C5f6 │ 0 ETH                    │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ Yes       │ 0xDCc3Af9Ad1e23047156e92621D8FFC7C1da3b6eC │ 0.198801497961625535 ETH │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ No        │ 0xE54a37226B0bCC5bE80485d63Ac12e18eC8212EA │ 0 ETH                    │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ No        │ 0xb3aD00CB3fEcbfe729cb496C70aC420FDc229AAC │ 4.0945156101 ETH         │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ Yes       │ 0xcC0678E51a8237b762c09d6548d2d07285609e98 │ 0.249559196 ETH          │ 0 NU │
├───────────┼────────────────────────────────────────────┼──────────────────────────┼──────┤
│ No        │ 0xe10deF5e87C34600b5Fa4d4f4D2aed74675F9A2F │ 0 ETH                    │ 0 NU │
╘═══════════╧════════════════════════════════════════════╧══════════════════════════╧══════╛
(nucypher) kieran@pegasus:~/Git/nucypher$ 

```

### CODE REVIEW

File: nucypher/cli/painting.py [hunk 1]
```diff
@@ -380,6 +382,57 @@ def paint_staking_confirmation(emitter, staker, new_stake):
...
+        if not active_stakes:
+            emitter.echo(f"There are no active stakes\n")
+
+        fees = staker.policy_agent.get_reward_amount(staker.checksum_address)
+        gwei_fees = f"{Web3.fromWei(fees, 'gwei')} Gwei"
+        last_confirmed = staker.staking_agent.get_last_active_period(staker.checksum_address)
+        missing = staker.staking_agent.get_missing_confirmations(staker_address=staker.checksum_address)
+
+        staker_data = [f'Missing {missing} confirmation{"s" if missing > 1 else ""}' if missing else f'Confirmed #{last_confirmed}',
+                       f'{"Yes" if staker.is_restaking else "No"} ({"Locked" if staker.restaking_lock_enabled else "Unlocked"})',
```
Comment (derekpierre) [inline nucypher/cli/painting.py]: We talked about potentially changing the term "Locked" vs "Unlocked"...I think it was more along the lines of modifiable/configurable or something similar - https://github.com/nucypher/nucypher/issues/1618.

File: nucypher/blockchain/eth/agents.py [hunk 2]
```diff
@@ -605,6 +605,22 @@ def get_completed_work(self, bidder_address: str):
...
 
+    @validate_checksum_address
+    def get_missing_confirmations(self, checksum_address: str) -> int:
+        # TODO: Move this up one layer, since it utilizes a combination of contract API methods.
+        last_confirmed_period = self.get_last_active_period(checksum_address)
+        current_period = self.get_current_period()
+        missing_confirmations = current_period - last_confirmed_period
+        if missing_confirmations in (0, -1):
+            result = 0
+        elif missing_confirmations == current_period:  # never confirmed
```
Comment (vzotova) [inline nucypher/blockchain/eth/agents.py]:
```suggestion
        elif last_confirmed_period == 0:  # never confirmed
```

File: nucypher/blockchain/eth/token.py [hunk 3]
```diff
@@ -296,6 +296,14 @@ def time_remaining(self, slang: bool = False) -> Union[int, str]:
             result = delta
         return result
 
+    def describe(self) -> Dict[str, str]:
+        start_datetime = self.start_datetime.local_datetime().strftime("%b %d %H:%M %Z")
+        data = dict(index=self.index,
+                    value=str(self.value),
+                    remaining=self.periods_remaining,
+                    enactment=start_datetime)
+        return data
```
Comment (vzotova) [inline nucypher/blockchain/eth/token.py]: I think will be useful to add here `final_period` datetime

File: nucypher/cli/commands/stake.py [hunk 4]
```diff
@@ -701,7 +703,7 @@ def prolong(general_config, transacting_staker_options, config_file, force, lock
     # Interactive
     if not lock_periods:
         stake_extension_range = click.IntRange(min=1, max=economics.maximum_allowed_locked, clamp=False)
-        max_extension = economics.maximum_allowed_locked - current_stake.periods_remaining
+        max_extension = economics.maximum_rewarded_periods - current_stake.periods_remaining
```
Comment (vzotova) [inline nucypher/cli/commands/stake.py]: it's better, but still is not fully correct, fixed in #1652

File: nucypher/cli/painting.py [hunk 5]
```diff
@@ -447,6 +448,57 @@ def paint_staking_confirmation(emitter, staker, new_stake):
     emitter.echo(next_steps, color='green')
 
 
+def paint_stakes(emitter, stakeholder, paint_inactive: bool = False, staker_address: str = None):
+    headers = ('Idx', 'Value', 'Remaining', 'Enactment')
```
Comment (vzotova) [inline nucypher/cli/painting.py]: last period will be helpful here, because it's hard to calculate this value from provided information

Comment (KPrasch) [reply nucypher/cli/painting.py]: :heavy_check_mark:

File: nucypher/cli/commands/stake.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py]: Since it's in an another open PR - Shall I do anything here?

File: nucypher/blockchain/eth/token.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/token.py]: :heavy_check_mark:

File: nucypher/blockchain/eth/agents.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/agents.py]: :heavy_check_mark:

File: nucypher/cli/commands/stake.py (hunk 4 above)
Comment (vzotova) [reply nucypher/cli/commands/stake.py]: I think it's ok to just leave it (if PR will be merged before) or discard - if merged after #1652

**Pull request merged**